Property refactor

jfirebaugh opened the pull request:
Ports `PropertyValue`, `TransitionablePropertyValue`, `PossiblyEvaluatedPropertyValue`, etc. from native.

Fixes #2739
Fixes #3044
Fixes #2769
Fixes #5627
Fixes not #3425 -- exception no longer occurs, but the layer turns blank. Probably a variant of #4012.

* [x] Fix CI
* [x] Fix `Paint` benchmark performance
* [x] Write some architecture docs
* [x] Test transitions
* ~[ ] Include fix for #4012?~ Nah, followup PR.
* ~[ ] Include fix for #3479?~ Nah, wait until the style/render split.

jfirebaugh commented:
The `Paint` benchmarks shows reduced performance, and it looks difficult to maintain the same performance while still fixing #2769. `StyleLayer::recalculate` obtains its current performance by essentially ignoring default values -- i.e. it's fast, but incorrect.

anandthakker commented on src/render/draw_symbol.js:
```diff
@@ -168,8 +166,7 @@ function drawTileSymbols(program, programConfiguration, painter, layer, tile, bu
     if (isSDF) {
-        const haloWidthProperty = `${isText ? 'text' : 'icon'}-halo-width`;
-        const hasHalo = !layer.isPaintValueFeatureConstant(haloWidthProperty) || layer.paint[haloWidthProperty];
+        const hasHalo = layer.paint.get(`${isText ? 'text' : 'icon'}-halo-width`).constantOr(1) !== 0;
```
I was gonna be pretty impressed that Flow could handle this, but it turns out it's just bailing -- adding typo to the property name expression here doesn't produce a flow error.   It does work if the conditional logic is used to select a string literal; i.e., `isText ? 'text-halo-width' : 'icon-halooooo-width'` produces an error, whereas ``${isText ? 'text' : 'icon'}-haloooo-width`` does not.

jfirebaugh replied:
👍 , will switch to `isText ? 'text-halo-width' : 'icon-halo-width'`.

anandthakker commented on src/shaders/describe_shaders:
```diff
@@ -49,7 +48,7 @@ style.layers.forEach((layer) => {
+    styleLayer.updatePaintTransitions([], {}, { zoom: zoom }, {});
     const configuration = ProgramConfiguration.createDynamic(
         programInterface, styleLayer, zoom);
     const key = `${layer.type}${configuration.cacheKey || ''}`;
```
I'd be okay with just removing `describe_shaders` -- it's not part of any regular workflows, and that makes it very likely to go stale

jfirebaugh replied:
👍

anandthakker commented on src/style/properties.js:
```diff
@@ -0,0 +1,645 @@
+            }
+        }
+        return false;
+    }
```
Along the lines of the point above re: the temporal state in `TransitioningPropertyValue`, would it make sense for `hasTransition()` to accept a `now` parameter (here and elsewhere)?

anandthakker commented on src/data/bucket/symbol_bucket.js:
```diff
@@ -414,13 +438,34 @@ class SymbolBucket implements Bucket {
-            this.textSizeData = getSizeData(this.zoom, layer, 'text-size');
-            this.iconSizeData = getSizeData(this.zoom, layer, 'icon-size');
+            const layer: SymbolStyleLayer = this.layers[0];
+            const unevaluatedLayoutValues = layer._unevaluatedLayout._values;
```
Oh, is this because `getLayoutProperty`'s return type isn't parameterized on the property name?

jfirebaugh replied:
It's because `layer.getLayoutProperty` returns the raw style spec value (as needed for the public runtime styling API), and we need the `PropertyValue`.

anandthakker replied:
Gotcha.  I guess we could add `SymbolStyleLayer#get{Text,Icon}Size()`... but allowing treating *_bucket as a 'friend' of *_style_layer is probably reasonable, too.

jfirebaugh commented on src/data/program_configuration.js:
```diff
@@ -90,28 +97,30 @@ class ConstantBinder implements Binder {
                        length: number,
                        feature: Feature) {
-        const value = layer.getPaintValue(this.property, {zoom: 0}, feature);
+        const value: any = this.expression.evaluate({zoom: 0}, feature);
```
Nope, this was just copypasta.

jfirebaugh commented on src/style/properties.js:
```diff
@@ -0,0 +1,645 @@
+export type PossiblyEvaluatedValue<T> =
+    | {kind: 'constant', value: T}
+    | SourceExpression
+    | CompositeExpression;
```
It's used only in `query_utils.js`, to quarantine an `any` cast. We could probably remove this with some additional refactoring.

jfirebaugh commented:
I figured out how to restore `StyleLayer::recalculate` performance using prototypal inheritance -- see followup commit. This puts benchmarks in good territory: no significant performance regressions, improvement in `LayerCircle` and `QueryPoint` (probably due to the possibly evaluated mechanism) and `StyleLayerCreate`.

https://bl.ocks.org/anonymous/raw/aad0ba1349e552599828bc7e5feadae4/

jfirebaugh commented on src/style/style.js:
```diff
@@ -307,25 +299,48 @@ class Style extends Evented {
+            }
+        }
+
+        // if (this.placement.hasTransition()) {
```
Need to uncomment and fix this. @ChrisLoer Is `this.placement.stillFading()` the right thing to check? It looks a bit different than [native's `hasTransitions()`](https://github.com/mapbox/mapbox-gl-native/blob/110c6a8de9942ac7090ef561ca1bed150b547150/src/mbgl/text/placement.cpp#L309-L311).

ChrisLoer replied:
Err.... hopefully we can improve this with a back-port from native. I think right now, the answer may be that you should ignore the placement state in this `hasTransitions`. They key logic for keeping animation running for placement/collision fading is the `needsRerender = !this.placement.isDone() || this.placement.stillFading()` part that sets `this._placementDirty` in `Map#_render`. That accounts for both collision fade animations (that's what `stillFading()` checks) but also for the map being in a state where it's waiting for the next placement to occur or in the middle of an incremental placement (kind of an analog of `stale` on the native side although it's not quite the same).

For now, if you just leave this out of `style.hasTransitions`, it should be fine because the render logic checks this state separately. /cc @ansis

mollymerp commented on src/style/properties.js:
```diff
@@ -631,6 +641,45 @@ class HeatmapColorProperty implements Property<Color, void> {
+ * on the default value: using `for (const property of Object.keys(this._values))`, they can iterate over
+ * only the _own_ properties of `_values`, skipping repeated calculation of transitions and possible/final
+ * evaluations for defaults, the result of which will always be the same.
+ *
```
This is a very cool solution 🥇

jfirebaugh commented on src/data/bucket/symbol_bucket.js:
```diff
@@ -365,8 +368,29 @@ class SymbolBucket implements Bucket {
+    //   2. `icon-size` at the zoom level of the bucket. Used to calculate a per-feature size for source `icon-size`
+    //       expressions.
+    //   3. `text-size` and `icon-size` at the zoom level of the bucket, plus one. Used to calculate collision boxes.
+    //   4. `text-size` at zoom level 18. Used for something line-symbol-placement-related.
```
We should, but nobody seems remember the details! Tracking this in https://github.com/mapbox/mapbox-gl-js/issues/5683.

jfirebaugh commented on src/style/properties.js:
```diff
@@ -266,11 +385,40 @@ class Layout<Properties: Object> {
+ *
+ * `PossiblyEvaluatedValue` represents the three possible outcomes of this step: if the input value was a constant or
+ * camera expression, then the "possibly evaluated" result is a constant value. Otherwise, the input value was either
+ * a source expression or a camera expression, and we must defer final evaluation until supplied a feature. We separate
```
Yes, good catch, will fix.

jfirebaugh commented on src/data/program_configuration.js:
```diff
@@ -43,12 +46,11 @@ function packColor(color: Color): [number, number] {
     ];
 }
 
```
👍, will port the docs from native.

jfirebaugh commented:
The remaining unaddressed review comments were @anandthakker's observations about the interior mutability of `TransitioningPropertyValue`'s `prior` member. On the one hand, I agree this is an oddity in what is otherwise an immutable class. On the other, it's currently pretty much a verbatim port from native, and I'm hesitant to start introducing differences to what has been working well on native. On balance, I think I'd like to keep this as-is for now. @anandthakker does that sound ok?
